synapse/lib/multislabseqn.py

```diff
@@ -0,0 +1,447 @@
+                os.unlink(optspath)
+            except FileNotFoundError:  # pragma: no cover
+                pass
+
+            shutil.rmtree(fn)
+
+            await asyncio.sleep(0)
+
+        self.firstindx = offs + 1
+        self._setFirstIndx(self.tailslab, offs + 1)
+
+        if ridx is not None:
+            del self._ranges[:ridx]
+
+    def _isTimeToRotate(self):
+        # Only the tailseqn is eligible for rotation
+        assert self.tailseqn
+        return self.tailseqn.size() >= self.idx_per_slab
+
+    async def _makeSlab(self, startidx: int) -> Tuple[s_lmdbslab.Slab, s_slabseqn.SlabSeqn]:
+
+        async with self._openlock:  # Avoid race in two tasks making the same slab
+
+            item = self._openslabs.get(startidx)
+            if item is not None:
+                item[0].incref()
+                return item
+
+            fn = self.slabFilename(self.dirn, startidx)
+
+            slab = await s_lmdbslab.Slab.anit(fn, **self.slabopts)
+            seqn = slab.getSeqn('nexuslog')
+
+            self._openslabs[startidx] = slab, seqn
+
+            def fini():
+                self._openslabs.pop(startidx, None)
+
+            slab.onfini(fini)
+
+            return slab, seqn
+
+    @contextlib.asynccontextmanager
+    async def _getSeqn(self, ridx: int) -> AsyncIterator[s_slabseqn.SlabSeqn]:
+        '''
+        Get the sequence corresponding to an index into self._ranges
+        '''
+        if ridx == len(self._ranges) - 1:
+            assert self.tailslab and self.tailseqn
+            slab, seqn = self.tailslab, self.tailseqn
+
+        elif ridx == self._cacheridx:
+            assert self._cacheslab and self._cacheseqn
+            slab, seqn = self._cacheslab, self._cacheseqn
+
+        else:
+            startidx = self._ranges[ridx]
+
+            self._cacheridx = None
+
+            if self._cacheslab is not None:
+                await self._cacheslab.fini()
+
+            slab, seqn = self._cacheslab, self._cacheseqn = await self._makeSlab(startidx)
+            self._cacheridx = ridx
+
+        slab.incref()
+        try:
+            yield seqn
+        finally:
+            await slab.fini()
+
+    async def add(self, item: Any, indx=None) -> int:
+        '''
+        Add a single item to the sequence.
+        '''
+        advances = True
+
+        if indx is not None:
+            if indx < self.firstindx:
+                raise s_exc.BadIndxValu(mesg=f'indx lower than first index in sequence {self.firstindx}')
+
+            if indx < self._ranges[-1]:
+                ridx = self._getRangeIndx(indx)
+                assert ridx is not None
+
+                async with self._getSeqn(ridx) as seqn:
+                    seqn.add(item, indx=indx)
+
+                return indx
+
+            if indx >= self.indx:
+                self.indx = indx
+            else:
+                advances = False
+        else:
+            indx = self.indx
+
+        if self._isTimeToRotate():
```

Comment:
Since this is tight-loop and perf-critcal, maybe we should unroll this from being a method ( and make seqn.size a property rather than an accessor method ) and just do the `size >=` check inline?

Reply:
updated